### PR TITLE
feat(errorhandling)! enhance exceptions with details for errortype, errorcode and parameters

### PR DESCRIPTION
**Pull request by ntruchsess**
## Description

Exceptions in Framework.ErrorHandling have been enhanced. They now may contain an error-type, an error-code, a message-template and multiple parameters. This allows structured logging of error-details and a more precise display of errors in the frontend (and optionally internationalization by including multi-language error-message-resources for specific errorcodes into the frontend). As those details are optional this change is fully backwards-compatible and api-endpoints may be adjusted on an 'as needed' basis.

## Why

error-messages being returned from api-endpoints are just application-identifier (often a fully qualified class-name) plus string-message. This cannot easily be parsed by the frontend to map to more detailed (and optionally multilingual) help-texts. Parameters of methods throwing exceptions are embedded within the message and therefore not available explicitly.

## Issue

N/A (CPLP-3090)

## Checklist

Please delete options that are not relevant.

- [X] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [X] I have performed a self-review of my own code
- [X] I have successfully tested my changes locally
- [X] I have added tests that prove my changes work
- [X] I have checked that new and existing tests pass locally with my changes

### CODE REVIEW

File: src/administration/Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 1]
```diff
@@ -82,7 +83,7 @@ private async Task<CompanyWithAddressData> GetCompanyWithAddressAsyncInternal(Gu
         var companyWithAddress = await _portalRepositories.GetInstance<IApplicationRepository>().GetCompanyUserRoleWithAddressUntrackedAsync(applicationId).ConfigureAwait(false);
         if (companyWithAddress == null)
         {
-            throw new NotFoundException($"applicationId {applicationId} not found");
+            throw NotFoundException.Create(AdministrationRegistrationErrors.APPLICATION_NOT_FOUND, new ErrorParameter[] { new("applicationId", applicationId.ToString()) });
```
Comment (Phil91) [inline src/administration/Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs:86]: I'd suggest changing the ToString to nameof. Should result in the same

```suggestion
            throw NotFoundException.Create(AdministrationRegistrationErrors.APPLICATION_NOT_FOUND, new ErrorParameter[] { new("applicationId", nameof(applicationId)) });
```

File: src/framework/Framework.ErrorHandling.Web/GeneralHttpErrorHandler.cs [hunk 2]
```diff
@@ -45,10 +47,11 @@ public class GeneralHttpErrorHandler
         { HttpStatusCode.InternalServerError, new MetaData("https://datatracker.ietf.org/doc/html/rfc7231#section-6.6.1", "The server encountered an unexpected condition.") }
     }.ToImmutableDictionary();
 
-    public GeneralHttpErrorHandler(RequestDelegate next, ILogger<GeneralHttpErrorHandler> logger)
+    public GeneralHttpErrorHandler(RequestDelegate next, ILogger<GeneralHttpErrorHandler> logger, IErrorMessageService? errorMessageService = null)
```
Comment (Phil91) [inline src/framework/Framework.ErrorHandling.Web/GeneralHttpErrorHandler.cs:50]: I guess the errorMessageService is currently nullable so we don't have to adjust all the code at once and will be changed in the future to a not nullable?

If that's the case let's maybe add an todo here

**Review by Phil91 — CHANGES_REQUESTED** [review]
One minor question open.
Besides of that looking really good.

Could you please comment the sonar findings - in my opinion that are false positives

File: src/framework/Framework.ErrorHandling.Web/GeneralHttpErrorHandler.cs (hunk 2 above)
Comment (ntruchsess) [reply src/framework/Framework.ErrorHandling.Web/GeneralHttpErrorHandler.cs]: yes, that is the reason. I did add a TODO as reminder

**Review by Phil91 — APPROVED** [review]

**Pull request merged**